common/logger/src/lib.rs

```diff
@@ -1,74 +1,426 @@
         $crate::log::error!(target: $crate::DEFAULT_TARGET, $($arg)+);
-    )
+    })
 }
 
 /// Define debug macro that specify libra as the target
 // TODO Remove historical crit from code base since it isn't supported in Rust Log.
 #[macro_export]
 macro_rules! debug {
-    ($($arg:tt)+) => (
+    ($($arg:tt)+) => ({
+        if $crate::struct_log_enabled!($crate::log::Level::Debug) {
+            $crate::struct_log!($($arg)+);
+        }
         $crate::log::debug!(target: $crate::DEFAULT_TARGET, $($arg)+);
-    )
+    })
 }
 
 /// Define  macro that specify libra as the target
 #[macro_export]
 macro_rules! error {
-    ($($arg:tt)+) => (
+    ($($arg:tt)+) => ({
+        if $crate::struct_log_enabled!($crate::log::Level::Error) {
+            $crate::struct_log!($($arg)+);
+        }
         $crate::log::error!(target: $crate::DEFAULT_TARGET, $($arg)+);
-    )
+    })
 }
 
 /// Define info macro that specify libra as the target
 #[macro_export]
 macro_rules! info {
-    ($($arg:tt)+) => (
+    ($($arg:tt)+) => ({
+        if $crate::struct_log_enabled!($crate::log::Level::Info) {
+            $crate::struct_log!($($arg)+);
+        }
         $crate::log::info!(target: $crate::DEFAULT_TARGET, $($arg)+);
-    )
+    })
 }
 
 /// Define trace macro that specify libra as the target
 #[macro_export]
 macro_rules! trace {
-    ($($arg:tt)+) => (
+    ($($arg:tt)+) => ({
+        if $crate::struct_log_enabled!($crate::log::Level::Trace) {
+            $crate::struct_log!($($arg)+);
+        }
         $crate::log::trace!(target: $crate::DEFAULT_TARGET, $($arg)+);
-    )
+    })
 }
 
 /// Define warn macro that specify libra as the target
 #[macro_export]
 macro_rules! warn {
-    ($($arg:tt)+) => (
+    ($($arg:tt)+) => ({
+        if $crate::struct_log_enabled!($crate::log::Level::Warn) {
+            $crate::struct_log!($($arg)+);
+        }
         $crate::log::warn!(target: $crate::DEFAULT_TARGET, $($arg)+);
-    )
+    })
+}
+
+#[macro_export]
+#[cfg(feature = "no_struct_log")]
+macro_rules! struct_log_enabled {
+    ($level:expr) => {
+        false
+    };
+}
+
+#[macro_export]
+#[cfg(not(feature = "no_struct_log"))]
+macro_rules! struct_log_enabled {
+    ($level:expr) => {
+        $crate::log::log_enabled!(target: $crate::DEFAULT_TARGET, $level)
+            && $crate::struct_logger_enabled()
+    };
+}
+
+#[macro_export]
+#[cfg(feature = "no_struct_log")]
+macro_rules! struct_log {
+    ($($arg:tt)+) => {};
+}
+
+#[macro_export]
+#[cfg(not(feature = "no_struct_log"))]
+macro_rules! struct_log {
+    ($($arg:tt)+) => {
+        let mut entry = $crate::StructuredLogEntry::new_unnamed();
+        $crate::format_struct_args_and_pattern!(entry, $($arg)+);
+        $crate::send_struct_log!(entry);
+    }
+}
+
+#[macro_export]
+macro_rules! send_struct_log {
+    ($entry:expr) => {
+        if $crate::struct_logger_enabled() {
+            let mut entry = $entry;
+            entry.timestamp($crate::Utc::now().format("%F %T").to_string());
```

Comment:
Why aren't we using machine readable timestamps for structured logs? Do we anticipate humans would be reading the json or in a  presenter?

Reply:
At least one tailer that I saw (fluentbit) support any timestamp format. If reader can parse human readable format, why make it un-readable?